File: matlab/noisedetector.m
function newthreshold = noisedetector(energy)
    persistent threshold;
    if (isempty(threshold))
        threshold = 1024;
    end
    persistent bin;
    if (isempty(bin))
        bin = zeros(128,1);
    end
    persistent point;
    if (isempty(point))
        point = zeros(128,1);
    end
    
	% noise detector
	% frame's energy
	% 4ms decay
	bin = bin*0.995;
	% update bin
	i = idivide(energy,256);
	if (i>128)
		i=128;
	end
	bin(i) = bin(i) + 160;
	% form area histogram
	for j=1:128
		point(j) = point(j) + sum(bin(1:j));
	end
	% noise absent?
	if (point(10) >= point(128)/4)
		candidate = 0;
	else
		% if noise present, find threshold candidate
		eightypct = 0.8*point(128);
		for j=10:128
			if (point(j)>=eightypct && point(j-1)<eightypct)
				candidate = j * 256; % energy quantum
				break
			end
		end
	end
	if (candidate>=threshold)
		decay = 0.95;
	else
		decay = 0.60653;
	end
	threshold = candidate + decay * (threshold - candidate);
	threshold = max([1024 threshold]);
    newthreshold = threshold ;
end

File: Ex22b/source/noisedetector.c
/*
 * noisedetector.c
 *
 *  Created on: Nov 29, 2017
 *      Author: phlav
 */


static unsigned energybin[128] = {0};
static unsigned cumulative[128] = {0};
static unsigned threshold = 1024;

unsigned noisedetector(unsigned energy)
{
	int i;
	unsigned candidate;
	float decay;
	// decay
	for (i=0; i<128; i++)
	{
		energybin[i] = energybin[i]*0.995;
	}
	i = energy/256;
	if (i>127)
		i=127;
	energybin[i] = energybin[i] + 160;
	// create cumulative histogram
	cumulative[0] = energybin[0];
	for (i=1; i<128; i++)
	{
		cumulative[i] = cumulative[i-1] + energybin[i];
	}
	// noise absent?
	if (cumulative[9] >= cumulative[127]/4)
	{
		candidate = 0;
	}
	else
	{
		// if noise present, find threshold candidate
		unsigned eightypct = 0.8*cumulative[127];
		for (i=9; i<128; i++)
		{
			if ((cumulative[i]>=eightypct) && (cumulative[i-1]<eightypct))
			{
				candidate = i * 256; // energy quantum
			}
			if (candidate>=threshold)
			{
				decay = 0.95;
			}
			else
			{
				decay = 0.60653;
			}
		}
	}
	threshold = candidate + decay * (threshold - candidate);
	if (threshold < 1024)
	{
		threshold = 1024;
	}
	return threshold;

}
